File: Top-Interview-Questions/Array-and-Strings-Medium-Collection/3-Longest-Substring-Without-Repeating-Characters.cpp
// Solution #1 Brute Force
// TC: O(n) length of string
// SC: O(1) preallocating only 256 spaces
// Runtime: 164 ms, faster than 10.91%
// Memory Usage: 11.7 MB, less than 18.72%
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        int n = s.length();
        
        if(!n)
            return 0;
        
        if(n == 1)
            return 1;
        
        return findLongestSubstr(s, n);
    }
    
    int findLongestSubstr(string s, int n){
        
        int longest = 0;
        int count;
        
        for(int i = 0; i < n; i++){
            
            count = 0;
            vector<bool> store(256, false);
        
            if(longest > n-i)
                break;
            
            for(int j = i; j < n; j++){
                
                if(store[s[j] - 0])
                    break;
                
                store[s[j] - 0] = true;
                count++;
            }
            
            longest = count > longest ? count : longest;
        }
        
        return longest;
    }
};

File: Top-Interview-Questions/Array-and-Strings-Medium-Collection/3-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 * Runtime: 108 ms, faster than 61.48% of JavaScript online submissions for Longest Substring Without Repeating Characters.
 * Memory Usage: 39.1 MB, less than 78.61% of JavaScript online submissions for Longest Substring Without Repeating Characters.
 */
var lengthOfLongestSubstring = function(s) {
    if (!s || s.length === 0) {
        return 0;
    }

    let count = 0, maxCount = 0, start = 0;
    let map = new Map();

    for (let index = 0; index < s.length; index++) {
        const char = s.charAt(index);

        if (!map.has(char)) {
            count++;
        } else {
            maxCount = Math.max(maxCount, count);

            start = map.get(char) > start ? map.get(char) : start;
            count = index - start;
        }

        map.set(char, index);
    }

    return Math.max(maxCount, count);
};